README.md
# openscribe
Openscribe is now hosted on [GitLab](https://gitlab.com/openscribe/openscribe)


debian/changelog
openscribe (1.3-1~bionic1) bionic; urgency=medium

  * Fixed package dependencies to also accept libsox3 instead of just libsox2

 -- Matt Pharoah <mtpharoah@gmail.com>  Fri, 25 May 2018 17:58:12 -0400

openscribe (1.3-0~bionic1) bionic; urgency=medium

  * Release for Ubuntu 18.04 Bionic Beaver
  * Switch from gksu to pkexec

 -- Matt Pharoah <mtpharoah@gmail.com>  Sun, 13 May 2018 21:16:49 -0400

openscribe (1.2-0~xenial1) xenial; urgency=medium

  * Fixed audio crackling in Ubuntu 16.04
  * Changed audio buffer settings to fix buffer underruns
  * Fixed freeze when opening OpenScribe with a command line argument
  * Fixed freeze when opening OpenScribe from the Open With dialog in a file browser
  * Fixed changelog window to have proper word wrapping

 -- Matt Pharoah <mtpharoah@gmail.com>  Sun, 12 Jun 2016 22:42:30 -0400

openscribe (1.1-0ubuntu3~xenial1) xenial; urgency=medium

  * Added gksu as a dependency
  * Updated gtkmm dependency version
  * Publish for Xenial series

 -- Matt Pharoah <mtpharoah@gmail.com>  Sun, 17 Apr 2016 13:38:43 -0400

openscribe (1.1-0ubuntu2~trusty1) trusty; urgency=low

  * Fixed dependency issue that prevented OpenScribe from being installed on systems with a newer version of udev
  * Changed default audio chunk size from 15ms to 25ms to prevent audio crackling on slower systems
  * Changed error message when unable to load an audio file to be more informative

 -- Matt Pharoah <mtpharoah@gmail.com>  Sat, 06 Dec 2014 17:06:24 -0500

openscribe (1.1-0ubuntu1~trusty1) trusty; urgency=low

  * All footpedals should now be supported!

 -- Matt Pharoah <mtpharoah@gmail.com>  Mon, 10 Nov 2014 14:18:52 -0500

openscribe (1.0-0ubuntu7~trusty2) trusty; urgency=low

  * Package SoX dependency changed from libsox1b to (libsox2 OR libsox1b)

 -- Matt Pharoah <mtpharoah@gmail.com>  Fri, 07 Nov 2014 17:12:30 -0500

openscribe (1.0-0ubuntu7~trusty1) trusty; urgency=low

  * Fixed missing #include that resulted in a build error on gcc 4.7 or newer
  * Publish for Trusty series

 -- Matt Pharoah <mtpharoah@gmail.com>  Sat, 25 Oct 2014 21:17:03 -0400

openscribe (1.0-0ubuntu6) precise; urgency=low

  * Fixed a potential stack overflow when slowing audio

 -- Matt Pharoah <mtpharoah@gmail.com>  Tue, 21 Oct 2014 16:48:39 -0400

openscribe (1.0-0ubuntu5) precise; urgency=low

  * Fixed minor memory leak
  * Improved UI of configuration window
  * Limit size of audio file name label to prevent main window resizing

 -- Matt Pharoah <mtpharoah@gmail.com>  Sun, 19 Oct 2014 16:04:49 -0400

openscribe (1.0-0ubuntu4) precise; urgency=low

  * Fixed freeze when changing options while an audio file is loaded

 -- Matt Pharoah <mtpharoah@gmail.com>  Fri, 17 Oct 2014 23:58:33 -0400

openscribe (1.0-0ubuntu3) precise; urgency=low

  * Fixed typo in previous patch (version should be Version)

 -- Matt Pharoah <mtpharoah@gmail.com>  Fri, 17 Oct 2014 22:55:36 -0400

openscribe (1.0-0ubuntu2) precise; urgency=low

  * Fixed Debian package to install openscribe.desktop
  * This should fix OpenScribe not being added to the application menu and as an audio player

 -- Matt Pharoah <mtpharoah@gmail.com>  Fri, 17 Oct 2014 22:31:55 -0400

openscribe (1.0-0ubuntu1) precise; urgency=low

  * Initial release

 -- Matt Pharoah <mtpharoah@gmail.com>  Thu, 16 Oct 2014 21:41:13 -0400


changelog.hpp
#ifndef CHANGELOG_HPP_
#define CHANGELOG_HPP_

#include <gtkmm/window.h>
#include <gtkmm/button.h>
#include <gtkmm/hvbox.h>
#include <gtkmm/scrolledwindow.h>
#include <gtkmm/label.h>

#include "version.hpp"
#include "windowList.hpp"

class ChangelogWindow : public Gtk::Window {
  private:
	Gtk::ScrolledWindow scrollArea;
	Gtk::Button okayButton;

	Gtk::VBox rootLayout;
	Gtk::VBox changeList;
	Gtk::HBox buttonLayout;
	
	Gtk::Label updateMessage;
	
	std::vector<Gtk::Label*> labels;
	std::vector<Gtk::HBox*> changes;
	
	INLINE void addChangelogSection(char const *version) {
		Glib::ustring versionString("<b>OpenScribe ");
		versionString += version;
		versionString += "</b>";
		
		Gtk::Label *spacer = new Gtk::Label(Glib::ustring(" "));
		changeList.pack_start(*spacer, false, false);
		labels.push_back(spacer);
		
		Gtk::Label *versionLabel = new Gtk::Label();
		versionLabel->set_markup(versionString);
		versionLabel->set_alignment(Gtk::ALIGN_START);
		changeList.pack_start(*versionLabel, false, false);
		labels.push_back(versionLabel);
	}
	
	INLINE void addChangelogEntry(char const *change) {
		Gtk::HBox *entry = new Gtk::HBox();
		changeList.pack_start(*entry, false, false);
		changes.push_back(entry);
		
		Gtk::Label *bullet = new Gtk::Label(Glib::ustring("  •  "));
		bullet->set_valign(Gtk::ALIGN_START);
		entry->pack_start(*bullet, false, false);
		labels.push_back(bullet);
		
		Gtk::Label *changeLabel = new Gtk::Label(Glib::ustring(change));
		changeLabel->set_line_wrap(true);
		changeLabel->set_alignment(Gtk::ALIGN_START);
		entry->pack_start(*changeLabel, true, true);
		labels.push_back(changeLabel);
	}

  public:
	INLINE ChangelogWindow(const Version &since) : Gtk::Window() {
		set_title("OpenScribe Changelog");
		resize(610, 500);
		set_border_width(3);

		okayButton.set_label(Glib::ustring("Okay"));
		okayButton.signal_clicked().connect(sigc::mem_fun(*this, &ChangelogWindow::hide));

		scrollArea.set_policy(Gtk::POLICY_NEVER, Gtk::POLICY_AUTOMATIC);
		
		updateMessage.set_text(Glib::ustring("You have updated to a new version of OpenScribe. Here are the changes since you last used OpenScribe:"));
		updateMessage.set_alignment(Gtk::ALIGN_START);
		updateMessage.set_line_wrap(true);
		changeList.pack_start(updateMessage, false, false);

		// put other versions here, with newer versions at the top

		/* Version 1.3.0 */
		if (since < Version(1,3,1)) {
			addChangelogSection("1.3.1");

			addChangelogEntry("Fixed package dependencies to also accept libsox3 instead of just libsox2");
		}
		
		/* Version 1.3.0 */
		if (since < Version(1,3,0)) {
			addChangelogSection("1.3.0");

			addChangelogEntry("Released for Ubuntu 18.04");
			addChangelogEntry("Switched from gksu to pkexec");
		}
		
		/* Version 1.2.0 */
		if (since < Version(1,2,0)) {
			addChangelogSection("1.2.0");

			addChangelogEntry("Fixed audio crackling in Ubuntu 16.04");
			addChangelogEntry("Changed audio buffer settings to fix buffer underruns");
			addChangelogEntry("Fixed freeze when opening OpenScribe with a command line argument");
			addChangelogEntry("Fixed freeze when opening OpenScribe from the Open With dialog in a file browser");
			addChangelogEntry("Fixed changelog window to have proper word wrapping");
		}

		/* Version 1.1.2 */
		if (since < Version(1,1,2)) {
			addChangelogSection("1.1.2");

			addChangelogEntry("Fixed dependency issue that prevented OpenScribe from being installed on systems with a newer version of udev");
			addChangelogEntry("Changed default audio latency from 15ms to 25ms to prevent audio crackling on slower systems");
			addChangelogEntry("Changed error message when unable to load an audio file to be more informative");
		}

		/* Version 1.1.0 */
		if (since < Version(1,1,0)) {
			addChangelogSection("1.1.0");

			addChangelogEntry("All footpedals should now be supported. If your footpedal does not work with OpenScribe, please e-mail the developer with a bug report.");
			addChangelogEntry("Updated help section");
		}

		/* Version 1.0.7 */
		if (since < Version(1,0,7)) {
			addChangelogSection("1.0.7");

			addChangelogEntry("Fixed an issue where OpenScribe would not compile when using gcc version 4.7 or later.");
			addChangelogEntry("OpenScribe is now published for Trusty and Utopic in addition to Precise.");
			addChangelogEntry("If you are using Precise (Ubuntu 12.04), this update will not affect you.");
		}

		/* Version 1.0.6 */
		if (since < Version(1,0,6)) {
			addChangelogSection("1.0.6");

			addChangelogEntry("Fixed a potential stack overflow when slowing down audio.");
		}

		/* Version 1.0.5 */
		if (since < Version(1,0,5)) {
			addChangelogSection("1.0.5");

			addChangelogEntry("Fixed a minor memory leak when closing the configuration window while multiple devices are connected.");
			addChangelogEntry("Added Revert, Cancel, Apply, and Okay buttons to configuration window.");
			addChangelogEntry("Closing the configuration window now asks if you want to save changes instead of always saving them. (No message is shown if you have not made any changes.)");
			addChangelogEntry("Name of audio file currently playing is ellipsized if it is too long. (This prevents the main window from being resized and adding ugly empty space.)");
		}

		/* Version 1.0.4 */
		if (since < Version(1,0,4)) {
			addChangelogSection("1.0.4");

			addChangelogEntry("Fixed a bug where applying changes to options while an audio file is loaded would result in a freeze.");
		}

		/* Version 1.0.3 */
		if (since < Version(1,0,3)) {
			addChangelogSection("1.0.3");

			addChangelogEntry("Fixed missing .desktop file in Debian package. OpenScribe should now be added to your application menu, and your file browser should now add OpenScribe to the list of programs that can open audio files.");
		}

		/* Version 1.0 (Updating from QTranscribe) */

		if (since < Version(1,0,0)) {
			addChangelogSection("1.0 (upgrading from QTranscribe 1.0.15)");

			addChangelogEntry("Switched GUI from QT to GTK+ 3.0");
			addChangelogEntry("Support for footpedals whose pedals are axes rather than buttons");
			addChangelogEntry("Support for saving the configuration of more than one footpedal");
			addChangelogEntry("OpenScribe allows both a footpedal and a hand control (or multiple footpedals) to be used at the same time");
			addChangelogEntry("Can now increase and decrease the slow speed with a footpedal");
			addChangelogEntry("Can now assign two actions to each footpedal. If a pedal's action is set to \"Modifier Pedal\", and it is depressed, other pedals will perform their second action");
			addChangelogEntry("Added help section");
			addChangelogEntry("Fixed memory leak caused by QT");
			addChangelogEntry("Fixed rare crash when skipping back slowed audio");
			addChangelogEntry("Fixed bug where, sometimes after opening a new dictation while another was already open, playback would skip to the beginning of the file every time the audio was paused.");
			addChangelogEntry("Fixed bug where audio position slider would not always update correctly");
			addChangelogEntry("Fast forward and rewind sound effect volume decreased");
			addChangelogEntry("New application icon");
			addChangelogEntry("Added in-application changelog");
			addChangelogEntry("Miscellaneous bugfixes and optimizations");
		}

		add(rootLayout);
			rootLayout.pack_start(scrollArea, true, true);
				scrollArea.add(changeList);
			rootLayout.pack_start(buttonLayout, false, false);
				buttonLayout.pack_end(okayButton, false, false);
		show_all_children();
	}
	
	virtual ~ChangelogWindow() {
		for (Gtk::Label *label : labels) delete label;
		for (Gtk::HBox *change : changes) delete change;
	}


};


#endif /* CHANGELOG_HPP_ */
